runtime/frame_descriptors.c

```diff
@@ -168,64 +189,204 @@ void caml_init_frame_descriptors(void)
+  while (1) {
+    d = atomic_load_relaxed(fds->descriptors + h);
+    if (d == e) {
+      atomic_store_relaxed(fds->descriptors + h, &dummy_descr);
+      /* placeholder that should not disturb caml_find_frame_descr */
+      return;
+    }
+    h = (h+1) & fds->mask;
   }
+}
 
-  caml_global_barrier_end(b);
+static void remove_frame_descriptors(
+  caml_frame_descrs * fds, void ** frametables, int ntables)
+{
+  intnat * frametable, len, decrease = 0;
+  frame_descr * descr;
+  caml_frametable_list ** previous;
+
+  caml_plat_lock(&frame_mutex);
+
+  for (int i = 0; i < ntables; i++) {
+    frametable = frametables[i];
+    len = *frametable;
+    descr = (frame_descr *)(frametable + 1);
+    for (intnat j = 0; j < len; j++) {
+      invalid_entry(fds, descr);
+      descr = next_frame_descr(descr);
+    }
+    decrease += len;
+  }
+
+  fds->num_descr -= decrease;
+
+  previous = &(fds->frametables);
+
+  iter_list(fds->frametables, current) {
+  resume:
+    for (int i = 0; i < ntables; i++) {
+      if (current->frametable == frametables[i]) {
+        *previous = current->next;
+        caml_stat_free(current);
+        ntables--;
+        if (ntables == 0) goto release;
+        current = *previous;
+        frametables[i] = frametables[ntables];
+        goto resume;
+      }
+    }
+    previous = &(current->next);
+  }
+
+ release:
+  caml_plat_unlock(&frame_mutex);
+
+  /* wait for all readers to finish their work */
+  /* this is to completely remove the worst ever "possible" scenario
+     where a reader got the old value of the frame_descr address,
+     is interrupted before checking the ret_addr field, to finally
+     access the field while the memory block has now been freed by
+     the caller of caml_unregister_frametables function... */
+  /* still, concurrent read / remove must be extremely rare in practice
+     so we should not be waiting too long */
+  SPIN_WAIT {
+    uintnat v = atomic_load_acquire(&frame_readers);
+    if (CAMLlikely(v == 0)) break;
+  }
 }
 
-void caml_register_frametables(void **table, int ntables) {
-  struct frametable_array frametables = { table, ntables };
-  do {} while (!caml_try_run_on_all_domains(
-                 &stw_register_frametables, &frametables, 0));
+void caml_unregister_frametables(void ** frametables, int ntables)
+{
+  remove_frame_descriptors(&current_frame_descrs, frametables, ntables);
 }
 
-caml_frame_descrs caml_get_frame_descrs(void)
+void caml_unregister_frametable(void * frametables)
 {
-  return current_frame_descrs;
+  caml_unregister_frametables(&frametables, 1);
 }
 
-frame_descr* caml_find_frame_descr(caml_frame_descrs fds, uintnat pc)
+caml_frame_descrs* caml_get_frame_descrs(void)
+{
+  return &current_frame_descrs;
+}
+
+frame_descr* caml_find_frame_descr(caml_frame_descrs * fds, uintnat pc)
 {
   frame_descr * d;
   uintnat h;
 
-  h = Hash_retaddr(pc, fds.mask);
+  atomic_thread_fence(memory_order_acquire);
+
+  atomic_fetch_add(&frame_readers, 1);
```

Comment:
I don't know if the cost of these extra atomic operations for readers is neglectible or not. It could be useful to run Sandmark benchmarks on this PR to get an idea.

`caml_find_frame_descr` is called in `stack_scan_frames` in fiber.c, which is used in a performance-critical way in the minor GC logic. In that function it is used in a loop on each frame of the call stack. Could we perform this reader-side locking outside the loop in `stack_scan_frames`, instead of at each iteration of the loop?

Side question: could we use a reader-writer lock on our frame-descriptors global, instead of the atomic-based approach proposed in this PR?

Reply:
I think the `atomic_thread_fence(memory_order_acquire)` can be put in `caml_get_frame_descrs` function (so outside the performance critical loop). Note that on some architecture (e.g. `x86`), `atomic_thread_fence` is a no-op so the test with Sandmark should be performed on another one.

I will try to find a better way for synchronization than the `atomic_fetch_add` and `atomic_fetch_sub`.

Reply:
Wouldn't a reader-writer lock do the job? Taking a reader lock in `caml_find_frame_descr` may be costly, but we can do it in `stack_scan_frames` around the loop. Then we could even go back to non-atomic operations.

Reply:
Maybe we can change `caml_get_frame_descrs` for `caml_open_frame_descrs` and add a `caml_close_frame_descr`. The good point is that all synchronization will be performed in one place. The cons is that we must ensure that the two functions are always called in pair (as any lock mechanism). I suppose we can implement a reader-writer lock using a `caml_plat_condition` ?

Reply:
What i had in mind was to add bindings to pthread's rwlocks to platform.h, in addition to the current locks we have bound there. We haven't used reader-writer locks in the runtime yet, but this is bound to happen someday.

(But maybe there is a reason why we need lower-level atomic operations instead of reader-writer locks here that I missed?)

> We can change `caml_get_frame_descrs` for `caml_open_frame_descrs` and add a `caml_close_frame_descrs`.

Yes, that would be a nice approach to exposing the read locking in the API.

Reply:
> (But maybe there is a reason why we need lower-level atomic operations instead of reader-writer locks here that I missed?)

I do not think so. Moreover, reader phase should be at the beginning of a garbage collection while writer would eventually happen at the end of the Gc or during normal execution. So I think the probability that readers and writers run in full concurrency is pretty low.

> What i had in mind was to add bindings to pthread's rwlocks to platform.h

Ok, I was not sure why it was not in the `platform.h` header yet (I thought it was not portable enough).

Reply:
Okay, then my proposal would be as follows:

- add rwlock to platform.h
- use it for frametables, with an API that lets consumer takes a lot for several consecutive reads (I'm happy with `caml_{open,close}_frame_descrs`

I think that the result would be simpler than the current PR¹, and it adds a new binding that could be useful for the rest of the runtime as well.

¹: I am also uneasy with the busy-waiting in the current implementation. Using rwlocks let us notify the OS of the dependency instead.

Reply:
I agree with this. I will do the changes as soon as possible.